step-functions/step-functions-docker-handler/src/main/java/step/functions/handler/ProxyMessageHandler.java

```diff
@@ -0,0 +1,121 @@
+package step.functions.handler;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import step.grid.GridImpl;
+import step.grid.ProxyGridServices;
+import step.grid.TokenWrapper;
+import step.grid.agent.handler.MessageHandler;
+import step.grid.agent.tokenpool.AgentTokenWrapper;
+import step.grid.agent.tokenpool.TokenReservationSession;
+import step.grid.client.GridClient;
+import step.grid.client.GridClientConfiguration;
+import step.grid.client.LocalGridClientImpl;
+import step.grid.filemanager.FileManagerClient;
+import step.grid.filemanager.FileVersionId;
+import step.grid.io.InputMessage;
+import step.grid.io.OutputMessage;
+
+import java.io.File;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ProxyMessageHandler implements MessageHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(ProxyMessageHandler.class);
+
+    // Use a static in order to have a singleton instance of the Grid.
+    // Unfortunately the agent doesn't expose any agent-wide session to store singleton objects
+    private static final ConcurrentHashMap<String, Object> gridMap = new ConcurrentHashMap<>();
+
+    // Constants
+    public static final String GRID = "grid";
+
+    // Properties passed from the controller via message properties
+    public static final String MESSAGE_HANDLER = "$proxy.messageHandler";
+    public static final String MESSAGE_HANDLER_FILE_ID = "$proxy.messageHandler.file.id";
+    public static final String MESSAGE_HANDLER_FILE_VERSION = "$proxy.messageHandler.file.version";
+
+    // AgentConf properties
+    public static final String AGENT_CONF_DOCKER_LOCALGRID_PORT = "docker.localgrid.port";
+
+    // Maximum time for the container to start
+    private static final long CONTAINER_STARTUP_OFFSET = 120_000;
+
+    @Override
+    public OutputMessage handle(AgentTokenWrapper agentTokenWrapper, InputMessage inputMessage) throws Exception {
+        Map<String, String> agentProperties = agentTokenWrapper.getProperties();
+        Map<String, String> messageProperties = inputMessage.getProperties();
+        long keywordTimeoutMs = inputMessage.getCallTimeout();
+        // TODO set timeout according to keyword execution timeout
+        final long noMatchExistsTimeout = (keywordTimeoutMs - CONTAINER_STARTUP_OFFSET) <= 0 ? keywordTimeoutMs - 30_000 : keywordTimeoutMs - CONTAINER_STARTUP_OFFSET;
```

Comment:
This expression can return negative values if keywordTimeoutMs < 30_000. It also doesn't match with the comment of the constant "Maximum time for the container to start". After careful reflection using  noMatchExistsTimeout = CONTAINER_STARTUP_OFFSET = 120_000 would still be the better option